### ISSUE TITLE
1920 special sem1 qn10

### ISSUE THREAD

**Issue by eseutella** [nus-cs2030/2021-s2#316]
# Summary
Put a summary of the question here

## Description 
I'm confused about the testcase that is highlighted in yellow. Why does it print 2345432 instead of 3454324 when it is using substring(1,8)?

## Suggested Answer (if any):
Post your answer or the suggested answer, if any. 

## Screenshots (if any): 
<img width="296" alt="Screenshot 2021-04-24 022026" src="https://user-images.githubusercontent.com/77688520/115947960-be38b280-a4fd-11eb-84c9-7510b6da928d.png">

**Comment by ARPspoofing** [comment]
It is because the get method is not being called, so it will take the original value. So in this case the function get() which you will write will take the new cached value. If not get is called it will just act as if nothing happened although the new value is in the cache alr

**Comment by ruthkangyr** [comment]
> It is because the get method is not being called, so it will take the original value. So in this case the function get() which you will write will take the new cached value. If not get is called it will just act as if nothing happened although the new value is in the cache alr

How should we implement the CachedString class in this case?

**Comment by clementyee303** [comment]
Hi, this is how I did it. For the `forEach` method, it seems to be taking in the original value (probably because `get` isn't explicitly called?), so I stored the original value separately for that use case.

![image](https://user-images.githubusercontent.com/75613825/115962574-a8071280-a54e-11eb-8da5-3672f7b89bc8.png)

**Comment by Brandon-Chiu** [comment]
Hi, I'm still confused by the explanation.  Why is it that the "get" method is not called inside the "map" method? Don't we need to supply the result of the "get" method into mapper.apply? 
e.g. mapper.apply(this.get())

**Comment by eugenechua94** [comment]
Hi, I have a slight different method which only consist of one variable in the class, and it might be slightly simpler than the provided answer:

![image](https://user-images.githubusercontent.com/80498187/116220567-614c2f00-a77f-11eb-8a30-a4ede4d86706.png)

@clementyee303 
If you're talking about the flatMap method, as it takes in a string and returns another cachedString from mapper.apply(this.get()), another get() should be joined outside of mapper.apply() inorder to extract the string from the flatmapped cachedString to input it as a supplier string for return cachedString. Hope my explanation somewhat makes sense.

**Comment by clementyee303** [comment]
> Hi, I have a slightly different method which only consist of one variable in the class, and it might be slightly less complicated than the provided answer:
> 
> ![image](https://user-images.githubusercontent.com/80498187/116220567-614c2f00-a77f-11eb-8a30-a4ede4d86706.png)
> 
> @clementyee303
> If you're talking about the flatMap method, as it takes in a string and returns another cachedString from mapper.apply(this.get()), another get() should be joined outside of mapper.apply() inorder to extract the string from the flatmapped cachedString to input it as a supplier string for return cachedString. Hope my explanation somewhat makes sense.

Hi, does your implementation pass the test case as noted in the question? I was referring to the forEach method, not the flatMap method. The two attributes were implemented for the forEach method example

**Comment by eugenechua94** [comment]
@clementyee303 

Hi apologies, realised one of the test case wasn't right, my bad.

**Comment by rljw** [comment]
Hi, I'm still not sure why the test case prints 2345432 instead of 3454324, could anyone explain?

**Comment by eseutella** [comment]
> Hi, I'm still not sure why the test case prints 2345432 instead of 3454324, could anyone explain?

This is because the value will only be updated if you call get(). Hence if you refer to Clement's implementation above, he stored the old value under stored and the new value under storedMaps. His forEach() returns the old value while his get() returns the new value. In the test case, get() is not called hence the old value 2345432 will be called by forEach() and will be printed instead of the new value 3454324.

**Comment by LongJiAn99** [comment]
> Hi, this is how I did it. For the `forEach` method, it seems to be taking in the original value (probably because `get` isn't explicitly called?), so I stored the original value separately for that use case.
> 
> ![image](https://user-images.githubusercontent.com/75613825/115962574-a8071280-a54e-11eb-8da5-3672f7b89bc8.png)

Hi! Looking at this code I am not very sure whether it encapsulates the idea of caching. Being a CachedString, I think the code can be modified better by adding a cache attribute in the class itself which can then just be called when using the get() method instead of having the supplier run the get() method again each time it is being called. 
![image](https://user-images.githubusercontent.com/77440060/116340982-ab322500-a812-11eb-8b1d-61064e1a421b.png)
This is the code I've written which only changes the cache once the get() is being called. Hope it helps!

**Comment by clementyee303** [comment]
> 
> 
> > Hi, this is how I did it. For the `forEach` method, it seems to be taking in the original value (probably because `get` isn't explicitly called?), so I stored the original value separately for that use case.
> > ![image](https://user-images.githubusercontent.com/75613825/115962574-a8071280-a54e-11eb-8da5-3672f7b89bc8.png)
> 
> Hi! Looking at this code I am not very sure whether it encapsulates the idea of caching. Being a CachedString, I think the code can be modified better by adding a cache attribute in the class itself which can then just be called when using the get() method instead of having the supplier run the get() method again each time it is being called.
> ![image](https://user-images.githubusercontent.com/77440060/116340982-ab322500-a812-11eb-8b1d-61064e1a421b.png)
> This is the code I've written which only changes the cache once the get() is being called. Hope it helps!

Yeah, I did think about adding the caching functionality as per the lecture example but the question was a bit vague about it. Thanks for the better example :)
